### ISSUE TITLE
Adding `host_exports` (like `run_exports` except between `build` & `host` only)

### ISSUE THREAD

**Issue by jakirkham** [conda/ceps#77]
`run_exports` has been immensely helpful. Just to summarize it `run_exports`:

* Allows packages with knowledge of what they need to express it for easy downstream consumption
  * Packages that are part of the build can add dependency to `run`
    * `weak` only happens when the dependency is added to `host` (not `build`)
    * `strong` happens when a dependency is added to `build` or `host` 
  * Dependencies added can express ABI compatibility

However there are still some cases that are not covered. In particular there are cases where a dependency added to `build` needs to add itself to `host`, but there is no need to add it to `run` (in fact this would often be better to avoid). IOW `run_exports/strong` does not meet this need. Some examples include:

* Header-only dependencies of a package
  * https://github.com/conda/conda-build/issues/3796#issuecomment-549431589
* Tighter alignment in `build` & `host` of compiler dependencies (like OpenMP)
  * https://github.com/conda-forge/openmp-feedstock/issues/126
* Compiler runtime libraries that are statically linked (like `cudart`)
  * https://github.com/conda-forge/staged-recipes/pull/21723#issuecomment-1405444862

Solving these cases would involve a `host_exports` that allows adding other dependencies to `host`. Could imagine to use cases for a package providing `host_exports`

* `host_exports/weak` (default)
  * Add dependencies to `host` if providing package is added to `host`
  * Solves the header-only dependency case above for a library
* `host_exports/strong` (opt-in)
  * Add dependencies to `host` if providing package is in `build` or `host`
  * Handles the compiler dependency alignment problem (with OpenMP)
  * Handles statically linked compiler libraries (like `cudart`)
  * Of most use for compiler use cases

**Comment by jakirkham** [comment]
cc @xhochy (for awareness)

**Comment by jakirkham** [comment]
@jaimergp @wolfv would be interested to hear your perspectives on this idea? 🙂

**Comment by jaimergp** [comment]
I'm generally supportive of the idea provided that it's useful for a wide range of cases. A quick look made me see it as a "couple of edge cases". For example, it's not clear to me what "Add dependencies to host if providing package is added to host" means; if the package is already in host, why do we need to export anything? Is it to constrain it further? I guess we can use  `pin_compatible` but we want to avoid that "annoyance" if the package in build already knows how tell host, right?

**Comment by jakirkham** [comment]
Thanks Jaime! 🙏

Am happy with limiting it to the case where a package is added to `requirements/build`

So that would mean there is no need for `host_exports/weak` and `host_exports/strong` could just be `host_exports`

Does that same more reasonable?

**Comment by wolfv** [comment]
Yeah I hate the "strong/weak" terminology as it hides what the run export actually does. And nobody really knows whether strong applies to host as well, or not. So I am very in favor of cleaning this up and come up with a flexible way of choosing the exports on a from-to basis.

**Comment by isuruf** [comment]
Overall I like the idea. I proposed the same thing when v1 was being developed, but there was no interest. I'm having a hard time finding the conversation though.

More things I would like the CEP to add are
```
host_to_constrains - to support `run_exports: weak_constrained`
build_to_constrains - to support `run_exports: host_constrained`
build_to_build - to support transitive exports
host_to_host - to support transitive exports
```

**Comment by h-vetinari** [comment]
I just realized that I forgot a very important angle here. I was coming just from the recipe side, but we need to persist those different requirement-exports in the package metadata, which would also require modification of the repodata schema (c.f. e.g. #108). That makes it less feasible to do this only for v1 recipes, in the sense that `conda{,-build}` would have to at least be able to digest a new format of repodata.

**Comment by isuruf** [comment]
>  I'm not clear on build_to_build and host_to_host - why not just add these as dependencies?

Because the constraints might not be strict. For eg: build 1 of `B=1.*` might depend on `A=1.*`  and build 2 of `B=1.*` might depend on `A=2.*`, but any downstream of `B` (call it `C`)needs to link to both `B=1.*` and the version of `A` that it was compiled against.

This is useful when the headers of `B` uses headers of `A`  and `A` is a direct dependency of `C` even though it may  not be obvious.

**Comment by h-vetinari** [comment]
First draft (without specification, but including everything Isuru suggested) in #129
